torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,186 @@
+#include <torch/csrc/jit/passes/frozen_graph_optimizations.h>
+#include <torch/csrc/jit/passes/remove_dropout.h>
+#include <torch/csrc/jit/passes/utils/optimization_pass.h>
+#include <torch/csrc/jit/runtime/graph_executor.h>
+#include <torch/csrc/utils/memory.h>
+#include <vector>
+#include "ATen/core/functional.h"
+#include "ATen/core/interned_strings.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+using Tensor = at::Tensor;
+
+class ConcatLinearLayers : public torch::jit::OptimizationPass {
+  using torch::jit::OptimizationPass::OptimizationPass;
+  void handleBlock(Block* b) override {
+    // We are using an ordered list so that one knows that
+    // one only needs to move arguments forward to be a valid move, not
+    // backwards.
+    std::unordered_map<Value*, std::list<Node*>> tensorUsers;
+
+    for (Node* n : b->nodes()) {
+      // Grouping together all linear layers that use the same Tensor for input
+      if (n->kind() != aten::linear) {
+        continue;
+      }
+      
+      // Punting on support for non-constant params for now.
+      // We really just need to know that the shape of the weights and biases are constant.
+      if(nonConstantParameters(n)){
+        continue;
+      }
+
+      Value* op_value = n->inputs().at(0);
+      auto weight = n->namedInput("weight");
+      auto bias = n->namedInput("bias");
+
+      if (!constant_as<Tensor>(weight).has_value() ||
+          !constant_as<Tensor>(bias).has_value()) {
+        continue;
+      }
+      if (weight->type() == NoneType::get() ||
+          bias->type() == NoneType::get()) {
+        continue;
+      }
+
+      if (tensorUsers.find(op_value) == tensorUsers.cend()) {
+        tensorUsers.insert({op_value, std::list<Node*>()});
+      }
+      tensorUsers.find(op_value)->second.push_back(n);
+    }
+
+    // Check the uses of a tensor to find ones that can be combined
+    for (auto& kv : tensorUsers) {
+      Value* op_value = kv.first;
+      auto& uses = kv.second;
+
+      while (!uses.empty()) {
+        auto it_first = uses.begin();
+        Node* base_node = *it_first;
+        uses.erase(it_first);
+
+        auto weight_type = constant_as<Tensor>(base_node->namedInput("weight"))->dtype();
+        auto bias_type = constant_as<Tensor>(base_node->namedInput("bias"))->dtype();
+        std::vector<Node*> matching_nodes;
+        matching_nodes.push_back(base_node);
+
+        // Now iterate over the rest of the users of the set to
+        // see if there is anything that we can coaleasce with.
+        for (auto it = uses.begin(); it != uses.end();) {
+          auto candidate_node = *it;
+          auto candidate_weight = constant_as<Tensor>(candidate_node->namedInput("weight"));
+          auto candidate_weight_type = candidate_weight->dtype();
+          auto candidate_bias = constant_as<Tensor>(candidate_node->namedInput("bias"));
+          auto candidate_bias_type = candidate_bias->dtype();
```

Comment:
We also need to be checking that the non-zero dims are equivalent (or all dims if we are not doing slicing logic)

Reply:
The optimization still works correctly if the non-zero dims are not equal, a thing that I intentionally programmed in. The unittest test this case.

Reply:
I don't think so ?
```
>>> inp = [torch.rand([1, 4]), torch.rand([0, 3])]
>>> torch.cat(inp)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: torch.cat(): Sizes of tensors must match except in dimension 0. Got 4 and 3 in dimension 1 (The offending index is 1)
>>> 
```

Reply:
```
inp = torch.rand([12, 5])
w = torch.rand([10, 5])
torch._C._nn.linear(inp, w)
w3 = torch.rand([10, 1])
torch._C._nn.linear(inp, w3)
```
throws the following error:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-59a7793913f5> in <module>
     13 torch._C._nn.linear(inp, w)
     14 w3 = torch.rand([10, 1])
---> 15 torch._C._nn.linear(inp, w3)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (12x5 and 1x10)
```

Does that mean that this case would already be handled, or can we not assume that this is a checked condition? 
Might be prudent to add a check anyways.

Reply:
i think we should check that the tensors can be batched together - that the weights non-0 dim are equal. there might be other cases where the program would execute successfully  but not be easily batched, such as with broadcasting